### PR TITLE
feat: adding column of table schema change

### LINKED ISSUE #7908: Implement replacing table plan in meta service.

### LINKED ISSUE #7910: Integrate `ADD COLUMN` on the frontend with meta service.

### PR DESCRIPTION
**Pull request by BugenZhao**
Signed-off-by: Bugen Zhao <i@bugenzhao.com>
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

This PR implements adding columns of table schema change on the meta service and compute node.

- We'll calculate a mapping from old columns to new columns for the table in the frontend, and the meta service will apply this to the output indices of new dispatchers. (#8094, #8142).
- We introduce a new barrier manager command of `ReplaceTable` which will generate `Update` mutation, sharing some logic with scaling.
- Add the logic of start/cancel/finish altering job in the catalog manager.
- Handle the fragment ID update for the MergeExecutor in the compute node.

Note: this PR should not be marked as user-facing, until #7906 is done.

Tracking issue: #6903.
Close #7908. Close #7910.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/meta/src/manager/catalog/fragment.rs [hunk 1]
```diff
@@ -262,6 +263,94 @@ where
...
+        assert_eq!(table_fragment.state(), State::Initial);
+        table_fragment.set_table_id(table_id);
+
+        // Directly set to `Created` and `Running` state.
+        table_fragment.set_state(State::Created);
+        table_fragment.update_actors_state(ActorState::Running);
+
+        table_fragments.insert(table_id, table_fragment.clone());
+
+        // Update downstream `Merge`s.
```
Comment (BugenZhao) [inline src/meta/src/manager/catalog/fragment.rs:297]: This part is too dirty. 🥵

File: src/meta/src/manager/catalog/fragment.rs [hunk 2]
```diff
@@ -842,18 +924,26 @@ where
     pub async fn get_downstream_chain_fragments(
         &self,
         table_id: TableId,
-    ) -> MetaResult<Vec<Fragment>> {
+    ) -> MetaResult<Vec<(DispatchStrategy, Fragment)>> {
```
Comment (BugenZhao) [inline src/meta/src/manager/catalog/fragment.rs:927]: We also need the output indices in the dispatchers (then map it with column index mapping), so `Fragment` is not sufficient here.

File: src/stream/src/executor/merge.rs [hunk 3]
```diff
@@ -147,18 +147,26 @@ impl MergeExecutor {
...
                         barrier.as_update_merge(self.actor_context.id, self.upstream_fragment_id)
                     {
-                        assert!(update.new_upstream_fragment_id.is_none());
+                        let new_upstream_fragment_id = update
+                            .new_upstream_fragment_id
+                            .unwrap_or(self.upstream_fragment_id);
+                        let added_upstream_actor_id = update.added_upstream_actor_id.clone();
+                        let removed_upstream_actor_id: HashSet<_> =
+                            if update.new_upstream_fragment_id.is_some() {
+                                select_all.upstream_actor_ids().iter().copied().collect()
```
Comment (BugenZhao) [inline src/stream/src/executor/merge.rs:156]: Remove all actors if fragment id is changed.

https://github.com/risingwavelabs/risingwave/blob/54a92ddd2befe51ce73905aaa2a721fca0b20f74/proto/stream_plan.proto#L47-L50

File: src/meta/src/manager/catalog/mod.rs [hunk 4]
```diff
@@ -1451,6 +1451,84 @@ where
...
+            tables.insert(table.id, table.clone());
+            commit_meta!(self, tables)?;
+
+            let version = self
+                .notify_frontend(Operation::Update, Info::Table(table.to_owned()))
+                .await;
+
+            Ok(version)
+        } else {
+            unreachable!("table must not exist and must be in altering procedure");
```
Comment (yezizp2012) [inline src/meta/src/manager/catalog/mod.rs:1509]:
```suggestion
            unreachable!("table must exist and be in altering procedure");
```

File: src/meta/src/rpc/service/ddl_service.rs [hunk 5]
```diff
@@ -498,14 +500,29 @@ where
 
         let mut stream_job = StreamingJob::Table(None, req.table.unwrap());
         let fragment_graph = req.fragment_graph.unwrap();
+        let table_col_index_mapping =
+            ColIndexMapping::from_protobuf(&req.table_col_index_mapping.unwrap());
 
-        let (_ctx, _table_fragments) = self
-            .prepare_replace_table(&mut stream_job, fragment_graph)
+        let (ctx, table_fragments) = self
```
Comment (yezizp2012) [inline src/meta/src/rpc/service/ddl_service.rs:506]: Moving these logic into `ddl_controller` part.

**Review by yezizp2012 — APPROVED** [review]
LSTM!!! Except some workaround, we can refactor them later. 🥵

File: src/meta/src/manager/catalog/fragment.rs [hunk 6]
```diff
@@ -262,6 +263,94 @@ where
...
+
+        assert!(merge_updates.is_empty());
+
+        // Commit changes and notify about the changes.
+        commit_meta!(self, table_fragments)?;
+
+        self.notify_fragment_mapping(&old_table_fragment, Operation::Delete)
+            .await;
+        self.notify_fragment_mapping(&table_fragment, Operation::Add)
+            .await;
```
Comment (fuyufjh) [inline src/meta/src/manager/catalog/fragment.rs:349]: Can this 2 lines be merged into 1 Update?

Comment (BugenZhao) [reply src/meta/src/manager/catalog/fragment.rs:349]: I guess we can't. By specifying `Update` we will update the vnode mapping for all fragment in the list that is used for scaling, while in this case, the fragment list is replaced and we should delete the entry of vnode mapping for old fragments.

**Pull request merged**